server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -493,8 +494,31 @@ private void updateCleanupKeyToCountMapOnCacheInsertion(CleanupKey cleanupKey) {
             cleanupKeyToCountMap.computeIfAbsent(shardId, k -> new HashMap<>()).merge(cleanupKey.readerCacheKeyId, 1, Integer::sum);
         }
 
-        private void updateCleanupKeyToCountMapOnCacheEviction(CleanupKey cleanupKey) {
-            if (stalenessThreshold == 0.0 || cleanupKey.entity == null) {
+        /**
+         * Updates the cleanupKeyToCountMap and staleKeysCount when a cache eviction occurs.
+         *
+         * <p>This method is called when an entry is evicted from the cache.
+         * It decrements the count of the entry in the cleanupKeyToCountMap.
+         * It also decrements the staleKeysCount only if the entry was accounted.
+         * If the count of the CleanupKey becomes zero, it removes the CleanupKey from the map.
+         *
+         * <p> We update the cleanupKeyToCountMap on every key removed from cache
+         * except for the keys removed with reason Replaced
+         *
+         * @param cleanupKey   the CleanupKey that has been evicted from the cache
+         * @param notification RemovalNotification of the cache entry evicted
+         */
+        private void updateStaleCountOnEntryRemoval(CleanupKey cleanupKey, RemovalNotification<Key, BytesReference> notification) {
+            if (notification.getRemovalReason() == RemovalReason.REPLACED) {
+                // The reason of the notification is REPLACED when a cache entry's value is updated, since replacing an entry
+                // does not affect the staleness count, we skip such notifications.
+                return;
+            }
+            if (cleanupKey.entity == null) {
+                // entity will only be null when the shard is closed/deleted
```

Comment:
I confirmed cleanupKey.entity will be null for **both** closed and deleted. 
Closing an Index would call `removeIndex` of IndicesService. 
In [removeIndex](https://github.com/opensearch-project/OpenSearch/blob/bef1faa5cb39971cdff91ab8f77dfbe16e760751/server/src/main/java/org/opensearch/indices/IndicesService.java#L1063-L1095) method, the `indexService` and `indices` map is recreated without closed & deleted shards. 
The cacheEntityLookup function is getting an `indexService` of the shard from the `indices` map, since it cannot find that in the map, it returns a null and hence an Optional.empty().

```java
this.indicesRequestCache = new IndicesRequestCache(settings, (shardId -> {
            IndexService indexService = this.indices.get(shardId.getIndex().getUUID());
            if (indexService == null) {
                return Optional.empty();
            }
            return Optional.of(new IndexShardCacheEntity(indexService.getShard(shardId.id())));
        }), cacheService, threadPool, clusterService);
```